File: day05/main.go
package main

import (
	"bufio"
	"errors"
	"fmt"
	"log"
	"os"
	"regexp"
	"strconv"
	"strings"
)

const (
	puzzleInput = "input.txt"
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatalln(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatalln(err)
		}
	}()

	var grid1 *Grid
	var grid2 *Grid

	var rows [][]byte

	modeGrid := true

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		if modeGrid {
			line := scanner.Bytes()
			if len(line) == 0 {
				g, err := newGridFromRows(rows)
				if err != nil {
					log.Fatalln(err)
				}
				grid1 = g
				grid2 = grid1.Clone()
				modeGrid = false
				continue
			}
			row, err := parseGridRow(line)
			if err != nil {
				log.Fatalln(err)
			}
			if len(row) != 0 {
				rows = append(rows, row)
			}
			continue
		}
		instr, err := parseInstrLine(scanner.Text())
		if err != nil {
			log.Fatalln(err)
		}
		grid1.processInstr1(instr)
		grid2.processInstr2(instr)
	}

	if err := scanner.Err(); err != nil {
		log.Fatalln(err)
	}

	fmt.Println("Part 1:", string(grid1.tops()))
	fmt.Println("Part 2:", string(grid2.tops()))
}

type (
	Grid struct {
		grid [][]byte
	}
)

func newGridFromRows(rows [][]byte) (*Grid, error) {
	h := len(rows)
	if h == 0 {
		return nil, errors.New("No rows")
	}
	w := len(rows[0])
	for _, i := range rows {
		if len(i) != w {
			return nil, errors.New("Mismatched rows")
		}
	}
	grid := make([][]byte, 0, w)
	for i := 0; i < w; i++ {
		col := make([]byte, 0, h)
		for j := 0; j < h; j++ {
			c := rows[h-j-1][i]
			if c == '.' {
				break
			}
			col = append(col, c)
		}
		grid = append(grid, col)
	}
	return &Grid{
		grid: grid,
	}, nil
}

func (g Grid) Clone() *Grid {
	grid := make([][]byte, 0, len(g.grid))
	for _, i := range g.grid {
		row := make([]byte, len(i))
		copy(row, i)
		grid = append(grid, row)
	}
	return &Grid{
		grid: grid,
	}
}

func (g *Grid) processInstr1(instr Instr) {
	for i := 0; i < instr.a; i++ {
		g.push(instr.c, g.pop(instr.b))
	}
}

func (g *Grid) processInstr2(instr Instr) {
	stack := make([]byte, 0, instr.a)
	for i := 0; i < instr.a; i++ {
		stack = append(stack, g.pop(instr.b))
	}
	for i, l := 0, instr.a; i < l; i++ {
		g.push(instr.c, stack[l-i-1])
	}
}

func (g *Grid) pop(col int) byte {
	l := len(g.grid[col]) - 1
	b := g.grid[col][l]
	g.grid[col] = g.grid[col][:l]
	return b
}

func (g *Grid) push(col int, b byte) {
	g.grid[col] = append(g.grid[col], b)
}

func (g *Grid) top(col int) byte {
	return g.grid[col][len(g.grid[col])-1]
}

func (g *Grid) tops() []byte {
	t := make([]byte, 0, len(g.grid))
	for i := range g.grid {
		t = append(t, g.top(i))
	}
	return t
}

func (g Grid) String() string {
	var b strings.Builder
	for n, i := range g.grid {
		b.WriteString(strconv.Itoa(n + 1))
		b.WriteString(": ")
		b.WriteString(string(i))
		b.WriteString("\n")
	}
	return b.String()
}

type (
	Instr struct {
		a, b, c int
	}
)

var (
	lineRegex = regexp.MustCompile(`^move (\d+) from (\d+) to (\d+)$`)
)

func parseInstrLine(line string) (Instr, error) {
	captures := lineRegex.FindStringSubmatch(line)
	if len(captures) == 0 {
		return Instr{}, errors.New("Invalid line")
	}
	a, err := strconv.Atoi(captures[1])
	if err != nil {
		return Instr{}, err
	}
	b, err := strconv.Atoi(captures[2])
	if err != nil {
		return Instr{}, err
	}
	c, err := strconv.Atoi(captures[3])
	if err != nil {
		return Instr{}, err
	}
	return Instr{
		a: a,
		b: b - 1,
		c: c - 1,
	}, nil
}

func parseGridRow(line []byte) ([]byte, error) {
	row := make([]byte, 0, (len(line)/4)+1)
	for i, l := 0, len(line)-1; i < l; {
		var c byte
		switch line[i] {
		case ' ':
			if line[i+1] != ' ' {
				return nil, nil
			}
			c = '.'
		case '[':
			c = line[i+1]
		default:
			return nil, errors.New("Invalid line")
		}
		row = append(row, c)
		i += 4
	}
	return row, nil
}

File: day05/src/main.rs
use lazy_static::lazy_static;
use regex::Regex;
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

type BoxError = Box<dyn std::error::Error>;
type BoxResult<T> = Result<T, BoxError>;

fn main() -> BoxResult<()> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let mut grid1 = Grid::empty();
    let mut grid2 = Grid::empty();

    let mut rows = Vec::new();

    let mut mode_grid = true;

    for line in reader.lines() {
        let line = line?;
        if mode_grid {
            if line.len() == 0 {
                grid1 = Grid::from_rows(&rows)?;
                grid2 = grid1.clone();
                mode_grid = false;
                continue;
            }
            let row = parse_grid_row(line.as_bytes())?;
            if row.len() != 0 {
                rows.push(row);
            }
            continue;
        }
        let instr = Instr::from_str(&line)?;
        grid1.process_instr_1(&instr)?;
        grid2.process_instr_2(&instr)?;
    }
    println!("Part 1: {}\nPart 2: {}", grid1.tops()?, grid2.tops()?);
    Ok(())
}

#[derive(Clone)]
struct Grid {
    grid: Vec<Vec<u8>>,
}

impl Grid {
    fn empty() -> Self {
        Self { grid: Vec::new() }
    }

    fn from_rows(rows: &[Vec<u8>]) -> BoxResult<Self> {
        let h = rows.len();
        if h == 0 {
            return Err("No rows".into());
        }
        let w = rows[0].len();
        for i in rows {
            if i.len() != w {
                return Err("Mismatched rows".into());
            }
        }
        let mut grid = Vec::with_capacity(w);
        for i in 0..w {
            let mut col = Vec::with_capacity(h);
            for j in 0..h {
                let c = rows[h - j - 1][i];
                if c == b'.' {
                    break;
                }
                col.push(c);
            }
            grid.push(col);
        }
        Ok(Self { grid })
    }

    fn process_instr_1(&mut self, &Instr(a, b, c): &Instr) -> BoxResult<()> {
        for _ in 0..a {
            let k = self.pop(b)?;
            self.push(c, k)
        }
        Ok(())
    }

    fn process_instr_2(&mut self, &Instr(a, b, c): &Instr) -> BoxResult<()> {
        let mut stack = Vec::with_capacity(a);
        for _ in 0..a {
            stack.push(self.pop(b)?);
        }
        for &i in stack.iter().rev() {
            self.push(c, i)
        }
        Ok(())
    }

    fn pop(&mut self, col: usize) -> BoxResult<u8> {
        self.grid[col].pop().ok_or("No more items".into())
    }

    fn push(&mut self, col: usize, b: u8) {
        self.grid[col].push(b)
    }

    fn peek(&self, col: usize) -> BoxResult<u8> {
        Ok(self.grid[col]
            .last()
            .ok_or::<BoxError>("No more items".into())?
            .clone())
    }

    fn tops(&self) -> BoxResult<String> {
        let mut s = String::with_capacity(self.grid.len());
        for i in 0..self.grid.len() {
            s.push(self.peek(i)? as char);
        }
        Ok(s)
    }
}

struct Instr(usize, usize, usize);

impl Instr {
    fn from_str(line: &str) -> BoxResult<Self> {
        lazy_static! {
            static ref RE: Regex = Regex::new(r"^move (\d+) from (\d+) to (\d+)$").unwrap();
        }
        let captures = RE.captures(line).ok_or("Invalid line")?;
        Ok(Self(
            captures.get(1).ok_or("Invalid line")?.as_str().parse()?,
            captures
                .get(2)
                .ok_or("Invalid line")?
                .as_str()
                .parse::<usize>()?
                - 1,
            captures
                .get(3)
                .ok_or("Invalid line")?
                .as_str()
                .parse::<usize>()?
                - 1,
        ))
    }
}

fn parse_grid_row(line: &[u8]) -> BoxResult<Vec<u8>> {
    let mut row = Vec::with_capacity((line.len() / 4) + 1);
    for part in line.chunks(4) {
        if part.len() < 3 {
            break;
        }
        match part {
            &[b' ', v, ..] => {
                if v != b' ' {
                    return Ok(Vec::new());
                } else {
                    row.push(b'.');
                }
            }
            &[b'[', v, ..] => row.push(v),
            _ => return Err("Invalid line".into()),
        }
    }
    Ok(row)
}
